### PR TITLE
Implement textDocument/codeAction

### PR DESCRIPTION
**Pull request by felixfbecker**
#251 to be merged first

@tomv564

### CODE REVIEW

**Comment by felixfbecker** [comment]
Rebased, @tomv564 does this work for you?

File: src/test/typescript-service-helpers.ts [hunk 1]
```diff
@@ -2400,47 +2393,43 @@ export function describeTypeScriptService(createService: TypeScriptServiceFactor
...
-							}
-						}]
+							}]
+						}
 					}
-				}
+				});
 			});
-		} as any);
-	} as any);
```
Comment (tomv564) [inline src/test/typescript-service-helpers.ts]: Ah, nice to drop these 👍

File: src/typescript-service.ts [hunk 2]
```diff
@@ -1023,92 +1037,104 @@ export class TypeScriptService {
...
+			}))
+			.startWith({ op: 'add', path: '', value: [] } as AddPatch);
 	}
 
+	/**
+	 * The workspace/executeCommand request is sent from the client to the server to trigger command
+	 * execution on the server. In most cases the server creates a WorkspaceEdit structure and
+	 * applies the changes to the workspace using the request workspace/applyEdit which is sent from
+	 * the server to the client.
+	 */
```
Comment (tomv564) [inline src/typescript-service.ts]: Great to see some documentation on how this works!

File: src/typescript-service.ts [hunk 3]
```diff
@@ -991,6 +1038,106 @@ export class TypeScriptService {
...
+						throw new Error(`Expected source file ${change.fileName} to exist in configuration`);
+					}
+					const uri = path2uri(this.root, change.fileName);
+					changes[uri] = change.textChanges.map(({ span, newText }): TextEdit => ({
+						range: {
+							start: ts.getLineAndCharacterOfPosition(sourceFile, span.start),
+							end: ts.getLineAndCharacterOfPosition(sourceFile, span.start + span.length)
+						},
+						newText
+					}));
```
Comment (tomv564) [inline src/typescript-service.ts]: So that's where all of `convertTextChange` went. Nice :)

File: src/typescript-service.ts [hunk 4]
```diff
@@ -991,6 +1038,106 @@ export class TypeScriptService {
...
+	 * applies the changes to the workspace using the request workspace/applyEdit which is sent from
+	 * the server to the client.
+	 */
+	workspaceExecuteCommand(params: ExecuteCommandParams, span = new Span()): Observable<OpPatch> {
+		switch (params.command) {
+			case 'codeFix':
+				if (!params.arguments || params.arguments.length < 1) {
+					return Observable.throw(new Error(`Command ${params.command} requires arguments`));
+				}
+				return this.executeCodeFixCommand(params.arguments, span);
```
Comment (tomv564) [inline src/typescript-service.ts]: I forgot to highlight this earlier in the PR, but throwing all of `arguments` into `this.executeCodeFixCommand(...` feels a bit dirty. Perhaps the "proper" way is to make sure the FileTextChanges array itself is _in_ `params.arguments[0]`. What do you think?

**Comment by tomv564** [comment]
I just checked out and built this branch and verified both Rename and the codeFix Code Action work

File: src/typescript-service.ts (hunk 4 above)
Comment (felixfbecker) [inline src/typescript-service.ts]: Why do you think it's dirty?

Comment (tomv564) [inline src/typescript-service.ts]: What if in the future - besides the array of FileTextChanges - we also need to pass some `Context c` from the CodeAction to `executeCodeFixCommand`.

If the FileTextChanges live in `params.arguments[0]`, then `params.arguments[1]` would be the logical place to put this Context parameter.

If we keep the FileTextChanges in _all of_ `params.arguments`, then where would this extra Context parameter go?

Comment (felixfbecker) [inline src/typescript-service.ts]: I would prefer the first parameter to be an object `{ context, fileTextChanges }` then.

Comment (tomv564) [inline src/typescript-service.ts]: Yes, that sounds good! 
I guess you are free to modify codeFix's use of Command arguments at any time without compatibility issues, so there is no need to change anything now.

**Pull request merged**